lib/srv/db/cloud/watchers/rds_proxy.go

```diff
@@ -0,0 +1,213 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watchers
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/rds"
+	"github.com/aws/aws-sdk-go/service/rds/rdsiface"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+)
+
+// rdsDBProxyFetcher retrieves RDS Proxies and their custom endpoints.
+type rdsDBProxyFetcher struct {
+	cfg rdsFetcherConfig
+	log logrus.FieldLogger
+}
+
+// newRDSDBProxyFetcher returns a new RDS Proxy fetcher instance.
+func newRDSDBProxyFetcher(config rdsFetcherConfig) (Fetcher, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &rdsDBProxyFetcher{
+		cfg: config,
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: "watch:rdsproxy",
+			"labels":        config.Labels,
+			"region":        config.Region,
+		}),
+	}, nil
+}
+
+// Get returns RDS Proxies and proxy endpoints matching the watcher's
+// selectors.
+func (f *rdsDBProxyFetcher) Get(ctx context.Context) (types.Databases, error) {
+	databases, err := f.getRDSProxyDatabases(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return filterDatabasesByLabels(databases, f.cfg.Labels, f.log), nil
+}
+
+// getRDSProxyDatabases returns a list of database resources representing RDS
+// Proxies and custom endpoints.
+func (f *rdsDBProxyFetcher) getRDSProxyDatabases(ctx context.Context) (types.Databases, error) {
+	// Get a list of all RDS Proxies. Each RDS Proxy has one "default"
+	// endpoint.
+	rdsProxies, err := getRDSProxies(ctx, f.cfg.RDS, common.MaxPages)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Get all RDS Proxy custom endpoints sorted by the name of the RDS Proxy
+	// that owns the custom endpoints.
+	customEndpointsByProxyName, err := getRDSProxyCustomEndpoints(ctx, f.cfg.RDS, common.MaxPages)
+	if err != nil {
+		f.log.Debugf("Failed to get RDS Proxy endpoints: %v.", err)
+	}
+
+	databases := types.Databases{}
+	for _, dbProxy := range rdsProxies {
+		// TODO(greedy52) RDS Proxy supports MS SQL Server but it requires MS
+		// SQL Server engine support for IAM auth.
+		//
+		// https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/rds-proxy-setup.html#rds-proxy-connecting-sqlserver
+		if aws.StringValue(dbProxy.EngineFamily) == rds.EngineFamilySqlserver {
+			f.log.Debugf("RDS Proxy %q with engine family %q is not supported yet. Skipping.", aws.StringValue(dbProxy.DBProxyName), aws.StringValue(dbProxy.EngineFamily))
+			continue
+		}
+		if !aws.BoolValue(dbProxy.RequireTLS) {
+			f.log.Debugf("RDS Proxy %q doesn't support TLS. Skipping.", aws.StringValue(dbProxy.DBProxyName))
+			continue
+		}
+
+		if !services.IsRDSProxyAvailable(dbProxy) {
```

Comment:
What would you say about adding a Health-ping-check call ensuring that proxy is reachable. This will help avoid the scenario where db agent is deployed outside VPC and has the right permission to discovery RDS proxy but connection will fail because db agent is not able to dial RDS Proxy VPC from outside

Reply:
I believe this applies to other database types too. I usually deploy Aurora, ElastiCache etc. in private subnets so they are not accessible from outside the VPC. 

I am not sure if "watchers" is the best place to do pings as it can block for a long time and delays other discoveries. I feel maybe better to have some kind of health check for every database after they are "registered". What do you think? A "health-check" feature will benefit non cloud-hosted databases as well.